### PR TITLE
Assignment9+10 @ Tony Chan

### PR DESCRIPTION
**Pull request by tnadalini**

### CODE REVIEW

**Comment by NicholasJohnsonDev** [comment]
I'll review this one.

File: Assignment/Assignment.Tests/PingProcessTests.cs [hunk 1]
```diff
@@ -1,152 +1,37 @@
...
-    public void StringBuilderAppendLine_InParallel_IsNotThreadSafe()
-    {
-        IEnumerable<int> numbers = Enumerable.Range(0, short.MaxValue);
-        System.Text.StringBuilder stringBuilder = new();
-        numbers.AsParallel().ForAll(item => stringBuilder.AppendLine(""));
-        int lineCount = stringBuilder.ToString().Split(Environment.NewLine).Length;
-        Assert.AreNotEqual(lineCount, numbers.Count()+1);
-    }
-
-    readonly string PingOutputLikeExpression = @"
```
Comment (NicholasJohnsonDev) [inline Assignment/Assignment.Tests/PingProcessTests.cs:131]: You are going to want to keep this string.

File: Assignment/Assignment.Tests/Assignment.Tests.csproj [hunk 2]
```diff
@@ -1,16 +1,19 @@
...
+    <IsPackable>false</IsPackable>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.11.0" />
+    <PackageReference Include="MSTest.TestAdapter" Version="2.2.7" />
+    <PackageReference Include="MSTest.TestFramework" Version="2.2.7" />
+    <PackageReference Include="coverlet.collector" Version="3.1.0" />
+    <ProjectReference Include="..\Assignment\Assignment.csproj" />
+  </ItemGroup>
```
Comment (NicholasJohnsonDev) [inline Assignment/Assignment.Tests/Assignment.Tests.csproj:16]: Weird choice to update to newer packages, but not the latest.

**Comment by NicholasJohnsonDev** [comment]
You are going to want to use the test file supplied and not start from scratch.

It is odd that you did not include the [Directory.Build.props](https://github.com/IntelliTect-Samples/EWU-CSCD371-2022-Winter/blob/Assignment9%2B10/Assignment/Directory.Build.props) file supplied with the assignment. This file is essentially a solution-level csproj file instead of just at the project level. The below code was provided in the assignment, completing the requirement "and enabled .NET analyzers for both projects"; but as you got rid of it, I had to it mark ❌. You should really put back all the fies that were in the assignment folder.
```
    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
    <EnableNETAnalyzers>true</EnableNETAnalyzers>
    <CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>
```

## Instructions

1. Implement `PingProcess`' `public Task<PingResult> RunTaskAsync(string hostNameOrAddress)` ❌
   - First implement `public void RunTaskAsync_Success()` test method to test `PingProcess.RunTaskAsync()` using `"localhost"`. ❌
   - Do NOT use async/await in this implementation. ❌
2. Implement `PingProcess`' `async public Task<PingResult> RunAsync(string hostNameOrAddress)` ❌
   - First implement the `public void RunAsync_UsingTaskReturn_Success()` test method to test `PingProcess.RunAsync()` using `"localhost"` **without** using async/await. ❌
   - Also implement the `async public Task RunAsync_UsingTpl_Success()` test method to test `PingProcess.RunAsync()` using `"localhost"` but this time **DO** using async/await. ❌
3. Add support for an optional cancellation token to the `PingProcess.RunAsync()` signature. ❌
   Inside the `PingProcess.RunAsync()` invoke the token's `ThrowIfCancellationRequested()` method so an exception is thrown. ❌
   Test that, when cancelled from the test method, the exception thrown is an `AggregateException` ❌✔ with a `TaskCanceledException` inner exception ❌ using the test methods `RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrapping` ❌and `RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrappingTaskCanceledException` ❌ respectively.
4. Complete/fix **AND test** `async public Task<PingResult> RunAsync(IEnumerable<string> hostNameOrAddresses, CancellationToken cancellationToken = default)` which executes ping for and array of hostNameOrAddresses (which can all be "localhost") **in parallel**, adding synchronization if needed. ❌
   NOTE:
      - The order of the items in the stdOutput is irrelevent and expected to be intermingled.
      - StdOutput must have all the ping output returned (no lines can be missing) even though intermingled. ❌
5. Implement **AND test** `public Task<int> RunLongRunningAsync(ProcessStartInfo startInfo, Action<string?>? progressOutput, Action<string?>? progressError, CancellationToken token)` using `Task.Factory.StartNew()` and invoking `RunProcessInternal` with a `TaskCreation` value of `TaskCreationOptions.LongRunning` and a `TaskScheduler` value of `TaskScheduler.Current`.
   NOTE: This method does **NOT** use `Task.Run`.

## Extra Credit

- Test and implement `PingProcess.RunAsync(System.IProgress<T> progess)` so that you can capture the output as it occurs rather than capturing the output only when the process finishes. ❌

## Fundamentals

- Ensure you enable:
  - nullable reference types is enabled ✔
  - net6 targeted ✔
  - C# 10.0 ✔
  - and enabled .NET analyzers for both projects ❌
- All of the above should be unit tested ❌
- Choose simplicity over complexity ✔

File: Assignment/Assignment.Tests/PingProcessTests.cs [hunk 3]
```diff
@@ -1,152 +1,37 @@
...
-    public void Run_CaptureStdOutput_Success()
-    {
-        PingResult result = Sut.Run("localhost");
-        AssertValidPingOutput(result);
-    }
-
-    [TestMethod]
-    public void RunTaskAsync_Success()
-    {
-        // Do NOT use async/await in this test.
```
Comment (whitneybolar) [inline Assignment/Assignment.Tests/PingProcessTests.cs:60]: Here's some help! This and RunAsync_UsingTaskReturn_Success() are written very similarly. Where result = task.Result; you just need to get PingResult result  and Task<PingResult> task.

File: Assignment/Assignment.Tests/PingProcessTests.cs [hunk 4]
```diff
@@ -1,152 +1,37 @@
...
-        PingResult result = default;
-        // Test Sut.RunAsync("localhost");
-        AssertValidPingOutput(result);
-    }
-
-    [TestMethod]
-#pragma warning disable CS1998 // Remove this
-    async public Task RunAsync_UsingTpl_Success()
-    {
```
Comment (whitneybolar) [inline Assignment/Assignment.Tests/PingProcessTests.cs:76]: This is also similar to RunAsync_UsingTaskReturn_Success() but you need await the task before result = task.Result.

File: Assignment/Assignment.Tests/PingProcessTests.cs [hunk 5]
```diff
@@ -1,152 +1,37 @@
...
-        AssertValidPingOutput(result);
-    }
-#pragma warning restore CS1998 // Remove this
-
-
-    [TestMethod]
-    [ExpectedException(typeof(AggregateException))]
-    public void RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrapping()
-    {
-
```
Comment (whitneybolar) [inline Assignment/Assignment.Tests/PingProcessTests.cs:90]: hint: start by making a new  CancellationTokenSource and use CancellationToken

**Review by whitneybolar — COMMENTED** [review]
Implement PingProcess' public Task<PingResult> RunTaskAsync(string hostNameOrAddress) ❌ This was highlighted in red, did it get taken out?
First implement public void RunTaskAsync_Success() test method to test PingProcess.RunTaskAsync() using "localhost". ❌
Do NOT use async/await in this implementation. ❌
Implement PingProcess' async public Task<PingResult> RunAsync(string hostNameOrAddress) ❌
First implement the public void RunAsync_UsingTaskReturn_Success() test method to test PingProcess.RunAsync() using "localhost" without using async/await. ❌
Also implement the async public Task RunAsync_UsingTpl_Success() test method to test PingProcess.RunAsync() using "localhost" but this time DO using async/await. ❌
Add support for an optional cancellation token to the PingProcess.RunAsync() signature. ❌ Inside the PingProcess.RunAsync() invoke the token's ThrowIfCancellationRequested() method so an exception is thrown. ❌ Test that, when cancelled from the test method, the exception thrown is an AggregateException ❌ with a TaskCanceledException inner exception ❌ using the test methods RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrapping ❌and RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrappingTaskCanceledException ❌ respectively.
Complete/fix AND test async public Task<PingResult> RunAsync(IEnumerable<string> hostNameOrAddresses, CancellationToken cancellationToken = default) which executes ping for and array of hostNameOrAddresses (which can all be "localhost") in parallel, adding synchronization if needed. ❌NOTE:
The order of the items in the stdOutput is irrelevent and expected to be intermingled.
StdOutput must have all the ping output returned (no lines can be missing) even though intermingled. ❌
Implement AND test public Task<int> RunLongRunningAsync(ProcessStartInfo startInfo, Action<string?>? progressOutput, Action<string?>? progressError, CancellationToken token) using Task.Factory.StartNew() and invoking RunProcessInternal with a TaskCreation value of TaskCreationOptions.LongRunning and a TaskScheduler value of TaskScheduler.Current. Returning a Task<PingResult> is also okay. NOTE: This method does NOT use Task.Run.
Extra Credit
Test and implement PingProcess.RunAsync(System.IProgress<T> progess) so that you can capture the output as it occurs rather than capturing the output only when the process finishes. ❌
Fundamentals
Ensure you enable:
nullable reference types is enabled ✔
net6 targeted ✔
C# 10.0 ✔
and enabled .NET analyzers for both projects ❌
All of the above should be unit tested ❌
Choose simplicity over complexity ❌

**Pull request closed without merging**